### PR TITLE
Add InferenceSession options and Provider to ORTModel

### PR DESCRIPTION
**Pull request by fxmarty**
`ORTModel.load_model` was not accepting custom session options, making it hard for example to use a given number of CPU cores with an `ORTModel` for inference. Indeed, onnxruntime obey neither `psutil.cpu_affinity` nor `taskset`.

This may close https://github.com/huggingface/optimum/issues/262

## Before submitting
- [x] Did you make sure to update the documentation with your changes?

### CODE REVIEW

**Review by philschmid — APPROVED** [review]
LGTM ✅ I left one comment on the session option if we should define a default value. 
@mfuntowicz what do you think?

File: optimum/onnxruntime/modeling_ort.py
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:189]: why did you add this method? For documentation reasons?
Nice idea with the `add_start_docstrings`!

File: optimum/onnxruntime/modeling_ort.py
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:279]: I think here we should define `self.provider` as well. That way a user can already access `model.provider` otherwise it wouldn't make much sense to add it in the `to` method.

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (philschmid) [inline optimum/onnxruntime/modeling_seq2seq.py:210]: same as for encoder why a list?

File: tests/onnxruntime/test_modeling.py
Comment (philschmid) [inline tests/onnxruntime/test_modeling.py:67]: are those tests run on any device or noly when there is a GPU available?

File: tests/onnxruntime/test_modeling.py
Comment (philschmid) [inline tests/onnxruntime/test_modeling.py:79]: does it make sense? What happens if the model hast cuda and tensorrt?

File: tests/onnxruntime/test_modeling.py
Comment (philschmid) [inline tests/onnxruntime/test_modeling.py:82]: I don't understand this one? should `model.provider` return the value which provider is used for inference?

File: tests/onnxruntime/test_modeling.py
Comment (philschmid) [inline tests/onnxruntime/test_modeling.py:119]: same as above

**Review by philschmid — CHANGES_REQUESTED** [review]
I added a few comments since some details are not clear yet to me.

File: optimum/onnxruntime/modeling_ort.py
Comment (fxmarty) [reply optimum/onnxruntime/modeling_ort.py:131]: Yes, it returns a list. Basically an InferenceSession can have several providers at the same time, by order of priority. So you could have `["TensorrtExecutionProvider", "CUDAExecutionProvider", "CPUExecutionProvider"]`, or `["CUDAExecutionProvider", "CPUExecutionProvider"]` for example. See https://onnxruntime.ai/docs/execution-providers/

File: optimum/onnxruntime/modeling_ort.py
Comment (fxmarty) [reply optimum/onnxruntime/modeling_ort.py:160]: Because it is a feature from onnxruntime. But maybe not used so much, I am not sure, so if you prefer to support only a single execution provider I can change that, let me know!

File: optimum/onnxruntime/modeling_ort.py
Comment (fxmarty) [reply optimum/onnxruntime/modeling_ort.py:189]: I was thinking it makes sense to have `from_pretrained` in the `ORTModel` section of the documentation. If in the future we extend these pipeline wrappers to other backends (say, fx or OpenVino), they may have custom parameters that the documentation from `OptimizedModel` will not capture. So It makes sense to me to have the doc for `from_pretrained` in the subclasses as well.

What do you think?

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (fxmarty) [reply optimum/onnxruntime/modeling_seq2seq.py:210]: see above

File: optimum/onnxruntime/modeling_ort.py
Comment (fxmarty) [reply optimum/onnxruntime/modeling_ort.py:279]: I define it in the `__init__` from `ORTModel`: https://github.com/fxmarty/optimum/blob/21f4d919b202561968258fb6e8dbce67130e957b/optimum/onnxruntime/modeling_ort.py#L98 . Does it sound good?

File: tests/onnxruntime/test_modeling.py
Comment (fxmarty) [reply tests/onnxruntime/test_modeling.py:79]: An error will be raised here: https://github.com/fxmarty/optimum/blob/21f4d919b202561968258fb6e8dbce67130e957b/optimum/onnxruntime/modeling_ort.py#L163-L167

File: tests/onnxruntime/test_modeling.py
Comment (fxmarty) [reply tests/onnxruntime/test_modeling.py:82]: Yes it is confusing. ONNX Runtime adds the `CPUExecutionProvider` as a default if it is not provided, see: https://github.com/microsoft/onnxruntime/blob/0d6cbc6e57c2ecd0722cefaa87d300f301864e3b/onnxruntime/core/session/inference_session.cc#L1280-L1287

File: optimum/onnxruntime/modeling_ort.py
Comment (fxmarty) [reply optimum/onnxruntime/modeling_ort.py:93]: Because to me `model` and `config` should be required arguments, it does not make sense to initialize an `ORTModel` without an InferenceSesson provided. What do you think?

File: tests/onnxruntime/test_modeling.py
Comment (fxmarty) [reply tests/onnxruntime/test_modeling.py:68]: Good catch, I use the decorator `@require_torch_gpu` now.

File: optimum/onnxruntime/modeling_ort.py
Comment (philschmid) [reply optimum/onnxruntime/modeling_ort.py:93]: Lets stick with what we are doing everywhere with named args

File: optimum/onnxruntime/modeling_ort.py
Comment (philschmid) [reply optimum/onnxruntime/modeling_ort.py:131]: cc @mfuntowicz on this

**Review by philschmid — APPROVED** [review]
LGTM, but let's get a review from @mfuntowicz and @echarlaix or @michaelbenayoun before merging. Especially, regarding the "providers" (list)

File: optimum/onnxruntime/modeling_ort.py
Comment (echarlaix) [inline optimum/onnxruntime/modeling_ort.py:181]: I would remove those lines

File: optimum/onnxruntime/modeling_ort.py
Comment (echarlaix) [reply optimum/onnxruntime/modeling_ort.py:181]: Currently we only support cpu and cuda so I don't think it makes a difference to give the possibility to give a list of execution providers with a priority order, so for clarity I would keep it as a string for the moment

**Review by echarlaix — APPROVED** [review]
LGTM ! 🔥

**Comment by fxmarty** [comment]
Great! In any case, even if `CUDAExecutionProvider` is solely passed, the execution providers from onnxruntime is still a list, `["CUDAExecutionProvider", "CPUExecutionProvider"]`, so to me it makes still sense to have a list.

**Pull request merged**
